app/src/main/java/de/westnordost/streetcomplete/quests/max_height/AddMaxHeight.java

```diff
@@ -0,0 +1,99 @@
+{
+
+	private final OverpassMapDataDao overpassServer;
+
+	@Inject public AddMaxHeight(OverpassMapDataDao overpassServer)
+	{
+		this.overpassServer = overpassServer;
+	}
+
+	/*the query does not exclude elements which are not accessible by the public,
+	because a maxheight sign will very probably be visible at the entrance or beginning*/
+	private static final String QUERY_RESTRICTIONS = "['maxheight'!~'.']['maxheight:physical'!~'.']";
+
+	private static final String ROADS = "(primary|secondary|tertiary|trunk)(_link)?|motorway|service|residential|unclassified|living_street";
+
+	@Override public boolean download(BoundingBox bbox, MapDataWithGeometryHandler handler)
+	{
+		return overpassServer.getAndHandleQuota(getOverpassQuery(bbox), handler);
+	}
+
+	private static String getOverpassQuery(BoundingBox bbox)
+	{
+		return OverpassQLUtil.getGlobalOverpassBBox(bbox) +
+			"(" +
+			" node['barrier'='height_restrictor']" +  QUERY_RESTRICTIONS + ";" +
+			" node['amenity'='parking_entrance']['parking'~'^(underground|multi-storey)$']" +  QUERY_RESTRICTIONS + ";" +
+			" way['highway'~'^(" + ROADS + ")$']['tunnel'~'^(yes|building_passage)$']" +  QUERY_RESTRICTIONS + ";" +
+			" way['highway'~'^(" + ROADS + ")$']['covered'='yes']" +  QUERY_RESTRICTIONS + ";" +
+			");" +
+			"out meta geom;";
+	}
+
+	@Override public AbstractQuestAnswerFragment createForm()
+	{
+		return new AddMaxHeightForm();
+	}
+
+	@Override public void applyAnswerTo(Bundle answer, StringMapChangesBuilder changes)
+	{
+		boolean hasNoSign = answer.getBoolean(AddMaxHeightForm.NO_SIGN);
+		String maxheight = answer.getString(AddMaxHeightForm.MAX_HEIGHT);
+		if(hasNoSign)
+		{
+			/*TODO: maxheight=default or maxheight=no_sign?
+			according to https://taginfo.openstreetmap.org/keys/maxheight#values maxheight=no_sign is only used 14 times,
+			while maxheight=default is used about 10K times*/
+			changes.add("maxheight", "default");
+		}
+		else
+		{
+			if (maxheight != null)
+			{
+				changes.add("maxheight", maxheight);
+			}
+		}
+	}
+
+	@Override public String getCommitMessage() { return "Add maximum height"; }
+	@Override public int getIcon() { return R.drawable.ic_quest_max_height; }
+	@Override public int getTitle(@NonNull Map<String, String> tags)
+	{
+		boolean isParkingEntrance = "parking_entrance".equals(tags.get("amenity"));
+		boolean isHeightRestrictor = "height_restrictor".equals(tags.get("barrier"));
+		boolean isTunnel = "yes".equals(tags.get("tunnel"));
+
+		if (isParkingEntrance) return R.string.quest_maxheight_parking_entrance_title;
+		if (isHeightRestrictor) return R.string.quest_maxheight_height_restrictor_title;
+		if (isTunnel) return R.string.quest_maxheight_tunnel_title;
+
+		return R.string.quest_maxheight_title;
+	}
+
+	@Override public int getTitle() { return R.string.quest_maxheight_title; }
+	@Override public int getDefaultDisabledMessage() { return 0; }
+	@Nullable @Override public Boolean isApplicableTo(Element element) { return null; }
```

Comment:
I think it is. But it would be a long chain of if-checks, I figure. You could use several tagfilters. I.e.... (pseudo code)

```
TagFilterExpression nodeFilter = parser.parse("nodes with (barrier=height_restrictor or amenity=parking_entrance and parking ~ underground|multi-storey) and !maxheight and !maxheight:physical");
TagFilterExpression wayFilter = parser.parse("ways with highway~(primary|secondary|tertiary|trunk)(_link)?|motorway|service|residential|unclassified|living_street and (covered=yes or tunnel~yes|building_passage) !maxheight and !maxheight:physical");
return nodeFilter.matches(element) || wayFilter.matches(element);
```

To be honest, with a little extension on the TagFilterExpression class (adding a method that outputs the TagFilterExpression without the full thing (bbox and out geom meta;) as Overpass QL, perhaps these filters could be used as well in the getOverpassQuery method.